ui/apps/platform/src/Containers/Collections/CollectionForm.tsx

```diff
@@ -23,17 +23,53 @@ import {
 import BreadcrumbItemLink from 'Components/BreadcrumbItemLink';
 import useSelectToggle from 'hooks/patternfly/useSelectToggle';
 import { collectionsBasePath } from 'routePaths';
-import { ResolvedCollectionResponse } from 'services/CollectionsService';
 import { CollectionPageAction } from './collections.utils';
 import RuleSelector from './RuleSelector';
 import CollectionAttacher from './CollectionAttacher';
 import CollectionResults from './CollectionResults';
+import { Collection, ScopedResourceSelector, SelectorEntityType } from './types';
+
+type FormStateReducerAction =
+    | { type: 'setName'; name: string }
+    | { type: 'setDescription'; description: string }
+    | { type: 'setRules'; entity: SelectorEntityType; selector: ScopedResourceSelector }
+    | { type: 'attachCollection'; collectionId: string }
+    | { type: 'detachCollection'; collectionId: string };
+
+function formStateReducer(state: Collection, payload: FormStateReducerAction): Collection {
+    switch (payload.type) {
+        case 'setName':
+            return { ...state, name: payload.name };
+        case 'setDescription':
+            return { ...state, name: payload.description };
+        case 'setRules': {
+            const selectorRules = { ...state.selectorRules };
+            selectorRules[payload.entity] = payload.selector;
+            return { ...state, selectorRules };
+        }
+        case 'attachCollection':
+            return {
+                ...state,
+                embeddedCollectionIds: state.embeddedCollectionIds.concat(payload.collectionId),
+            };
+        case 'detachCollection':
+            return {
+                ...state,
+                embeddedCollectionIds: state.embeddedCollectionIds.filter(
+                    (id) => id !== payload.collectionId
+                ),
+            };
+        default:
+            // Type safe fallback to ensure we don't miss any cases
+            return ((_: never) => _)(payload);
```

Comment:
This might be useful to extract to a utility, but where to put it? And what to name it? `exhaustive()`, `unreachable()`, `cannotBeCalled()`... ?